7227 refactor transformer and diffusion model unet

marksgraham opened the pull request:
Part of #7227  .

### Description

A few sentences describing the changes proposed in this pull request.

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

KumoLiu commented on monai/networks/blocks/attention.py:
```diff
@@ -0,0 +1,327 @@
+Rearrange, _ = optional_import("einops.layers.torch", name="Rearrange")
+
+
+class SABlock(nn.Module):
```
Does this move break anything? If somewhere else calls the SABlock in selfattention directly?

marksgraham replied:
no its just the transformerblock that imports it directly. the transformer block is then used by a few networks (ViT, SwinUNET)

KumoLiu replied:
May not break anything in monai but may affect users?
Perhaps we need add `from monai.networks.blocks.attention import SABlock` in selfattention and add a deperecated warning? What do you think?

marksgraham commented:
Hi @KumoLiu @ericspod 

This PR is nearly ready for review, but one problem I'm encountering is that having refactored the attention blocks (self-attention, cross-attention, and spatial attention) to use einops a lot of the tests now fail as einops is not available. We could get around this using  ` @skipUnless(has_einops, "Requires einops")` but this would mean disabling a lot of tests. I make it we would have to disable at least some tests in:

```tests_autoencoderkl
test_controlnet
test_controlnet_inferers
test_diffusion_inferer
test_diffusion_model_unet
test_latent_diffusion_inferer
test_self_attention
test_cross_attention
test_spatial_attention
test_spade_autoencoderkl
test_transformer
```

Any suggestions on how I should proceed here?

KumoLiu commented:
Hi @marksgraham,
Add `@skipUnless(has_einops, "Requires einops")` should be ok since einops is a requirement in `requirements-dev.txt`. All these tests will be tested in the ci.
https://github.com/Project-MONAI/MONAI/blob/dev/requirements-dev.txt#L35
https://github.com/Project-MONAI/MONAI/blob/dev/.github/workflows/pythonapp.yml#L49
Thanks.

marksgraham commented:
Sorry for such a large PR! The main changes are:

- Created a new cross-attention and spatial attention block
- Modified autoencoderkl and diffusion_model_unet to make use of these new blocks
- Updated the transformerblock so it can optionally do cross-attention, too

I've verified locally that the models changed can still load in checkpoints from monai-generative and they produce the same results

ericspod commented:
I see that we're not using the `xformers` package now, I don't remember if we discussed why but what's the reason? 

We're removing the `use_flash_attention` in arguments for various classes, if they're all private that's fine (which I think they are) but any that weren't should have them back with deprecation warnings.

marksgraham commented:
> I see that we're not using the `xformers` package now, I don't remember if we discussed why but what's the reason?
> 
> We're removing the `use_flash_attention` in arguments for various classes, if they're all private that's fine (which I think they are) but any that weren't should have them back with deprecation warnings.

It was only included for flash attention, so removed as we are not using flash attention. There are two reasons I removed the xformers-based flash attention:
1) I never found it to find the memory savings it promised, and I think neither did Walter
2) There is now native support for flash attention in Pytorch [here](https://pytorch.org/docs/2.2/generated/torch.nn.functional.scaled_dot_product_attention.html) which we should look to integrate, but I think its only in 2.1 onwards for now

ericspod commented:
> 2\. There is now native support for flash attention in Pytorch [here](https://pytorch.org/docs/2.2/generated/torch.nn.functional.scaled_dot_product_attention.html) which we should look to integrate, but I think its only in 2.1 onwards for now

It's also still beta but let's keep an eye on it, thanks.

marksgraham commented on monai/networks/blocks/attention.py:
```diff
@@ -0,0 +1,327 @@
+        save_attn: bool = False,
+        causal: bool = False,
+        sequence_length: int | None = None,
+        rel_pos_embedding: Optional[str] = None,
```
agreed but think it best to leave new features to post-refactor

ericspod reviewed (approved):
I think this looks good now but would it be possible to have tests for `load_old_state_dict` defined for a number of networks? This would require creating a small network with the old code and storing that, that would be infeasible if the weight file is too large  but it's worth trying if we can. @KumoLiu is on national holiday but will be back shortly and can double check this one.

marksgraham commented:
> I think this looks good now but would it be possible to have tests for `load_old_state_dict` defined for a number of networks? This would require creating a small network with the old code and storing that, that would be infeasible if the weight file is too large but it's worth trying if we can. @KumoLiu is on national holiday but will be back shortly and can double check this one.

I think that's a good idea - how/to where can i upload small weights files for the testing?

atbenmurray commented:
> > I think this looks good now but would it be possible to have tests for `load_old_state_dict` defined for a number of networks? This would require creating a small network with the old code and storing that, that would be infeasible if the weight file is too large but it's worth trying if we can. @KumoLiu is on national holiday but will be back shortly and can double check this one.
> 
> I think that's a good idea - how/to where can i upload small weights files for the testing?

For testing I would suggest just generating the weight tensors on the fly

ericspod commented:
> weight tensors on the fly

The problem here is that we want to store the weights from networks defined in GenerativeModels, that is old network versions. I don't think we should add that repo as a testing prerequisite.

KumoLiu commented:
> I think that's a good idea - how/to where can i upload small weights files for the testing?

Hi @marksgraham, If the test weight isn't too large, you can add it [here](https://github.com/Project-MONAI/MONAI-extra-test-data/releases); if not, you might need to upload it to Google Drive.

Thanks,
Yun

KumoLiu commented on monai/networks/blocks/selfattention.py:
```diff
@@ -34,46 +34,76 @@ def __init__(
             num_heads (int): number of attention heads.
             dropout_rate (float, optional): fraction of the input units to drop. Defaults to 0.0.
+            hidden_input_size (int, optional): dimension of the input tensor. Defaults to hidden_size.
+            dim_head (int, optional): dimension of each head. Defaults to hidden_size // num_heads.
```
Looks like the change in this [PR](https://github.com/Project-MONAI/MONAI/pull/7664) already included in this one right?

marksgraham replied:
Yes unfortunately I realised that some of the changes in that PR were actually already existing in the Generative self-attention block and needed to be incorporated here in order to fully test everything here.

Apologies @NabJa for duplicating some work you had already done :(

NabJa replied:
You said we should wait with my PR and then copied some of my code. This is not the way it should be done. You could have at least merged my working branch into yours. My changes are minor and would not have caused complex merge conflicts.

KumoLiu replied:
Perhaps we can merge that one first since this refactor PR will be merged into `gen-ai-dev` branch, if there is any conflict we can solve it here?
Do you see any other concern for this proposal? @ericspod @marksgraham @NabJa

marksgraham replied:
Hi,

Yes my plan was always to ask @NabJa to merge their PR in, I'm happy to merge his in first and then solve conflicts here @KumoLiu .

@NabJa the reason I adopted much of the code from your PR was because I was hoping to reduce the merge conflicts when we did merge yours in. Maybe merging your branch in here would have been the better way to do it, I'll think about that next time :)

KumoLiu commented on monai/networks/blocks/selfattention.py:
```diff
@@ -34,46 +34,76 @@ def __init__(
+            dim_head (int, optional): dimension of each head. Defaults to hidden_size // num_heads.
             qkv_bias (bool, optional): bias term for the qkv linear layer. Defaults to False.
+            save_attn (bool, optional): to make accessible the attention matrix. Defaults to False.
+            causal: whether to use causal attention.
```
Can we add a reference here?

KumoLiu commented on monai/networks/blocks/crossattention.py:
```diff
@@ -0,0 +1,166 @@
+                For now only "decomposed" is supported (see https://arxiv.org/abs/2112.01526). 2D and 3D are supported.
+            input_size (tuple(spatial_dim), optional): Input resolution for calculating the relative
+                positional parameter size.
+            upcast_attention: if True, upcast attention operations to full precision.
```
May I ask why we need this `upcast_attention` here? If we want to cast the dtype, why not directly using `dtype` here?

marksgraham replied:
will update

KumoLiu commented on monai/networks/nets/diffusion_model_unet.py:
```diff
@@ -59,122 +55,9 @@ def zero_module(module: nn.Module) -> nn.Module:
-    https://github.com/Project-MONAI/MONAI/issues/7227
-
-    A basic Transformer block.
+    A Transformer block that allows for the input dimension to differ from the hidden dimension.
```
Is there a reference for this block?

marksgraham replied:
Not that i am aware of, I am afraid

KumoLiu commented on monai/networks/nets/diffusion_model_unet.py:
```diff
@@ -541,68 +296,19 @@ def forward(self, x: torch.Tensor, emb: torch.Tensor | None = None) -> torch.Ten
-            )
-
     def forward(self, x: torch.Tensor, emb: torch.Tensor | None = None) -> torch.Tensor:
         del emb
```
`emb` here is directly been deleted, is it expected? If yes, why not directly remove this arg?

KumoLiu reviewed (approved):
Overall looks good to me, minor comments inline.
And agreed with @ericspod, would be better to add test for load_old_state_dict.

Thanks.

marksgraham commented:
Merging in dev has created a big diff again. At least this PR has already been reviewed, so it shouldn't be too much of a problem for that.

I discussed with Eric and if needs be when the time comes to merge `gen-ai-dev` to `dev` we can create a new PR that only cherry picks the actual changes we have made on `gen-ai-dev` if we want to keep the commit history cleaner.

marksgraham commented:
@KumoLiu is there anything else that needs doing before this can be merged? Once it is merged I can finish the last bits of the refactor..!

KumoLiu commented:
/build

KumoLiu commented:
/build

KumoLiu commented:
blossom ci failed in test_diffusion_model_unet:
```[2024-05-10T07:22:05.391Z] ======================================================================
[2024-05-10T07:22:05.391Z] ERROR: test_shape_with_different_in_channel_out_channel (tests.test_diffusion_model_unet.TestDiffusionModelUNet3D)
[2024-05-10T07:22:05.391Z] ----------------------------------------------------------------------
[2024-05-10T07:22:05.391Z] Traceback (most recent call last):
[2024-05-10T07:22:05.391Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_diffusion_model_unet.py", line 518, in test_shape_with_different_in_channel_out_channel
[2024-05-10T07:22:05.391Z]     result = net.forward(torch.rand((1, in_channels, 16, 16, 16)), torch.randint(0, 1000, (1,)).long())
[2024-05-10T07:22:05.391Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/diffusion_model_unet.py", line 1672, in forward
[2024-05-10T07:22:05.391Z]     h = self.middle_block(hidden_states=h, temb=emb, context=context)
[2024-05-10T07:22:05.391Z]   File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T07:22:05.391Z]     return forward_call(*input, **kwargs)
[2024-05-10T07:22:05.391Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/diffusion_model_unet.py", line 793, in forward
[2024-05-10T07:22:05.391Z]     hidden_states = self.attention(hidden_states)
[2024-05-10T07:22:05.391Z]   File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T07:22:05.391Z]     return forward_call(*input, **kwargs)
[2024-05-10T07:22:05.391Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/blocks/spatialattention.py", line 76, in forward
[2024-05-10T07:22:05.391Z]     x = self.norm(x)
[2024-05-10T07:22:05.391Z]   File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T07:22:05.391Z]     return forward_call(*input, **kwargs)
[2024-05-10T07:22:05.391Z]   File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/normalization.py", line 268, in forward
[2024-05-10T07:22:05.391Z]     return F.group_norm(
[2024-05-10T07:22:05.391Z]   File "/opt/conda/lib/python3.8/site-packages/torch/nn/functional.py", line 2363, in group_norm
[2024-05-10T07:22:05.391Z]     return torch.group_norm(input, num_groups, weight, bias, eps, torch.backends.cudnn.enabled)
[2024-05-10T07:22:05.391Z] RuntimeError: Unsupported memory format. Supports only ChannelsLast, Contiguous
```
cc @marksgraham

marksgraham commented:
@KumoLiu which pytorch version was this on?

KumoLiu commented:
> @KumoLiu which pytorch version was this on?

Pytorch version: 1.10.0a0+0aef44c in `nvcr.io/nvidia/pytorch:21.10-py3`

marksgraham commented:
@KumoLiu i think i've fixed it. one of the outstanding issues on the refactor is to investigate why these `.contiguous()` calls are needed in `DiffusionModelUNet` and to see if they can be removed so I'll look into it more properly soon

KumoLiu commented:
/build

KumoLiu commented:
Still failed.
<details><summary>Error Log</summary>
<p>

```
[2024-05-10T11:03:33.714Z] 
[2024-05-10T11:03:33.714Z] ======================================================================
[2024-05-10T11:03:33.714Z] ERROR: test_shape_unconditioned_models_0 (tests.test_spade_diffusion_model_unet.TestDiffusionModelUNet3D)
[2024-05-10T11:03:33.714Z] ----------------------------------------------------------------------
[2024-05-10T11:03:33.714Z] Traceback (most recent call last):
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/parameterized/parameterized.py", line 620, in standalone_func
[2024-05-10T11:03:33.714Z]     return func(*(a + p.args), **p.kwargs, **kw)
[2024-05-10T11:03:33.714Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_spade_diffusion_model_unet.py", line 519, in test_shape_unconditioned_models
[2024-05-10T11:03:33.714Z]     result = net.forward(
[2024-05-10T11:03:33.714Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/spade_diffusion_model_unet.py", line 932, in forward
[2024-05-10T11:03:33.714Z]     h = upsample_block(hidden_states=h, res_hidden_states_list=res_samples, seg=seg, temb=emb, context=context)
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.714Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.714Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/spade_diffusion_model_unet.py", line 304, in forward
[2024-05-10T11:03:33.714Z]     hidden_states = resnet(hidden_states, temb, seg)
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.714Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.714Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/spade_diffusion_model_unet.py", line 169, in forward
[2024-05-10T11:03:33.714Z]     h = self.norm1(h, seg)
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.714Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.714Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/blocks/spade_norm.py", line 88, in forward
[2024-05-10T11:03:33.714Z]     normalized = self.param_free_norm(x)
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.714Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/container.py", line 141, in forward
[2024-05-10T11:03:33.714Z]     input = module(input)
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.714Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/normalization.py", line 268, in forward
[2024-05-10T11:03:33.714Z]     return F.group_norm(
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/functional.py", line 2360, in group_norm
[2024-05-10T11:03:33.714Z]     return torch.group_norm(input, num_groups, weight, bias, eps, torch.backends.cudnn.enabled)
[2024-05-10T11:03:33.714Z] RuntimeError: Unsupported memory format. Supports only ChannelsLast, Contiguous
[2024-05-10T11:03:33.714Z] 
[2024-05-10T11:03:33.714Z] ======================================================================
[2024-05-10T11:03:33.714Z] ERROR: test_shape_unconditioned_models_1 (tests.test_spade_diffusion_model_unet.TestDiffusionModelUNet3D)
[2024-05-10T11:03:33.714Z] ----------------------------------------------------------------------
[2024-05-10T11:03:33.714Z] Traceback (most recent call last):
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/parameterized/parameterized.py", line 620, in standalone_func
[2024-05-10T11:03:33.714Z]     return func(*(a + p.args), **p.kwargs, **kw)
[2024-05-10T11:03:33.714Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_spade_diffusion_model_unet.py", line 519, in test_shape_unconditioned_models
[2024-05-10T11:03:33.714Z]     result = net.forward(
[2024-05-10T11:03:33.714Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/spade_diffusion_model_unet.py", line 932, in forward
[2024-05-10T11:03:33.714Z]     h = upsample_block(hidden_states=h, res_hidden_states_list=res_samples, seg=seg, temb=emb, context=context)
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.714Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.714Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/spade_diffusion_model_unet.py", line 304, in forward
[2024-05-10T11:03:33.714Z]     hidden_states = resnet(hidden_states, temb, seg)
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.714Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.714Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/spade_diffusion_model_unet.py", line 169, in forward
[2024-05-10T11:03:33.714Z]     h = self.norm1(h, seg)
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.714Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.714Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/blocks/spade_norm.py", line 88, in forward
[2024-05-10T11:03:33.714Z]     normalized = self.param_free_norm(x)
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.714Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/container.py", line 141, in forward
[2024-05-10T11:03:33.714Z]     input = module(input)
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.714Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/normalization.py", line 268, in forward
[2024-05-10T11:03:33.714Z]     return F.group_norm(
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/functional.py", line 2360, in group_norm
[2024-05-10T11:03:33.714Z]     return torch.group_norm(input, num_groups, weight, bias, eps, torch.backends.cudnn.enabled)
[2024-05-10T11:03:33.714Z] RuntimeError: Unsupported memory format. Supports only ChannelsLast, Contiguous
[2024-05-10T11:03:33.714Z] 
[2024-05-10T11:03:33.714Z] ======================================================================
[2024-05-10T11:03:33.714Z] ERROR: test_shape_unconditioned_models_2 (tests.test_spade_diffusion_model_unet.TestDiffusionModelUNet3D)
[2024-05-10T11:03:33.714Z] ----------------------------------------------------------------------
[2024-05-10T11:03:33.714Z] Traceback (most recent call last):
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/parameterized/parameterized.py", line 620, in standalone_func
[2024-05-10T11:03:33.714Z]     return func(*(a + p.args), **p.kwargs, **kw)
[2024-05-10T11:03:33.714Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_spade_diffusion_model_unet.py", line 519, in test_shape_unconditioned_models
[2024-05-10T11:03:33.714Z]     result = net.forward(
[2024-05-10T11:03:33.714Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/spade_diffusion_model_unet.py", line 932, in forward
[2024-05-10T11:03:33.714Z]     h = upsample_block(hidden_states=h, res_hidden_states_list=res_samples, seg=seg, temb=emb, context=context)
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.714Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.714Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/spade_diffusion_model_unet.py", line 304, in forward
[2024-05-10T11:03:33.714Z]     hidden_states = resnet(hidden_states, temb, seg)
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.714Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.714Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/spade_diffusion_model_unet.py", line 169, in forward
[2024-05-10T11:03:33.714Z]     h = self.norm1(h, seg)
[2024-05-10T11:03:33.714Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.714Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.714Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/blocks/spade_norm.py", line 88, in forward
[2024-05-10T11:03:33.715Z]     normalized = self.param_free_norm(x)
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.715Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/container.py", line 141, in forward
[2024-05-10T11:03:33.715Z]     input = module(input)
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.715Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/normalization.py", line 268, in forward
[2024-05-10T11:03:33.715Z]     return F.group_norm(
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/functional.py", line 2360, in group_norm
[2024-05-10T11:03:33.715Z]     return torch.group_norm(input, num_groups, weight, bias, eps, torch.backends.cudnn.enabled)
[2024-05-10T11:03:33.715Z] RuntimeError: Unsupported memory format. Supports only ChannelsLast, Contiguous
[2024-05-10T11:03:33.715Z] 
[2024-05-10T11:03:33.715Z] ======================================================================
[2024-05-10T11:03:33.715Z] ERROR: test_shape_unconditioned_models_3 (tests.test_spade_diffusion_model_unet.TestDiffusionModelUNet3D)
[2024-05-10T11:03:33.715Z] ----------------------------------------------------------------------
[2024-05-10T11:03:33.715Z] Traceback (most recent call last):
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/parameterized/parameterized.py", line 620, in standalone_func
[2024-05-10T11:03:33.715Z]     return func(*(a + p.args), **p.kwargs, **kw)
[2024-05-10T11:03:33.715Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_spade_diffusion_model_unet.py", line 519, in test_shape_unconditioned_models
[2024-05-10T11:03:33.715Z]     result = net.forward(
[2024-05-10T11:03:33.715Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/spade_diffusion_model_unet.py", line 922, in forward
[2024-05-10T11:03:33.715Z]     h = self.middle_block(hidden_states=h, temb=emb, context=context)
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.715Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.715Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/diffusion_model_unet.py", line 793, in forward
[2024-05-10T11:03:33.715Z]     hidden_states = self.attention(hidden_states)
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.715Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.715Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/blocks/spatialattention.py", line 76, in forward
[2024-05-10T11:03:33.715Z]     x = self.norm(x)
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.715Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/normalization.py", line 268, in forward
[2024-05-10T11:03:33.715Z]     return F.group_norm(
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/functional.py", line 2360, in group_norm
[2024-05-10T11:03:33.715Z]     return torch.group_norm(input, num_groups, weight, bias, eps, torch.backends.cudnn.enabled)
[2024-05-10T11:03:33.715Z] RuntimeError: Unsupported memory format. Supports only ChannelsLast, Contiguous
[2024-05-10T11:03:33.715Z] 
[2024-05-10T11:03:33.715Z] ======================================================================
[2024-05-10T11:03:33.715Z] ERROR: test_shape_unconditioned_models_4 (tests.test_spade_diffusion_model_unet.TestDiffusionModelUNet3D)
[2024-05-10T11:03:33.715Z] ----------------------------------------------------------------------
[2024-05-10T11:03:33.715Z] Traceback (most recent call last):
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/parameterized/parameterized.py", line 620, in standalone_func
[2024-05-10T11:03:33.715Z]     return func(*(a + p.args), **p.kwargs, **kw)
[2024-05-10T11:03:33.715Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_spade_diffusion_model_unet.py", line 519, in test_shape_unconditioned_models
[2024-05-10T11:03:33.715Z]     result = net.forward(
[2024-05-10T11:03:33.715Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/spade_diffusion_model_unet.py", line 922, in forward
[2024-05-10T11:03:33.715Z]     h = self.middle_block(hidden_states=h, temb=emb, context=context)
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.715Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.715Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/diffusion_model_unet.py", line 793, in forward
[2024-05-10T11:03:33.715Z]     hidden_states = self.attention(hidden_states)
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.715Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.715Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/blocks/spatialattention.py", line 76, in forward
[2024-05-10T11:03:33.715Z]     x = self.norm(x)
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.715Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/normalization.py", line 268, in forward
[2024-05-10T11:03:33.715Z]     return F.group_norm(
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/functional.py", line 2360, in group_norm
[2024-05-10T11:03:33.715Z]     return torch.group_norm(input, num_groups, weight, bias, eps, torch.backends.cudnn.enabled)
[2024-05-10T11:03:33.715Z] RuntimeError: Unsupported memory format. Supports only ChannelsLast, Contiguous
[2024-05-10T11:03:33.715Z] 
[2024-05-10T11:03:33.715Z] ======================================================================
[2024-05-10T11:03:33.715Z] ERROR: test_shape_unconditioned_models_5 (tests.test_spade_diffusion_model_unet.TestDiffusionModelUNet3D)
[2024-05-10T11:03:33.715Z] ----------------------------------------------------------------------
[2024-05-10T11:03:33.715Z] Traceback (most recent call last):
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/parameterized/parameterized.py", line 620, in standalone_func
[2024-05-10T11:03:33.715Z]     return func(*(a + p.args), **p.kwargs, **kw)
[2024-05-10T11:03:33.715Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_spade_diffusion_model_unet.py", line 519, in test_shape_unconditioned_models
[2024-05-10T11:03:33.715Z]     result = net.forward(
[2024-05-10T11:03:33.715Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/spade_diffusion_model_unet.py", line 922, in forward
[2024-05-10T11:03:33.715Z]     h = self.middle_block(hidden_states=h, temb=emb, context=context)
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.715Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.715Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/diffusion_model_unet.py", line 793, in forward
[2024-05-10T11:03:33.715Z]     hidden_states = self.attention(hidden_states)
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.715Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.715Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/blocks/spatialattention.py", line 76, in forward
[2024-05-10T11:03:33.715Z]     x = self.norm(x)
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.715Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/normalization.py", line 268, in forward
[2024-05-10T11:03:33.715Z]     return F.group_norm(
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/functional.py", line 2360, in group_norm
[2024-05-10T11:03:33.715Z]     return torch.group_norm(input, num_groups, weight, bias, eps, torch.backends.cudnn.enabled)
[2024-05-10T11:03:33.715Z] RuntimeError: Unsupported memory format. Supports only ChannelsLast, Contiguous
[2024-05-10T11:03:33.715Z] 
[2024-05-10T11:03:33.715Z] ======================================================================
[2024-05-10T11:03:33.715Z] ERROR: test_shape_unconditioned_models_6 (tests.test_spade_diffusion_model_unet.TestDiffusionModelUNet3D)
[2024-05-10T11:03:33.715Z] ----------------------------------------------------------------------
[2024-05-10T11:03:33.715Z] Traceback (most recent call last):
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/parameterized/parameterized.py", line 620, in standalone_func
[2024-05-10T11:03:33.715Z]     return func(*(a + p.args), **p.kwargs, **kw)
[2024-05-10T11:03:33.715Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_spade_diffusion_model_unet.py", line 519, in test_shape_unconditioned_models
[2024-05-10T11:03:33.715Z]     result = net.forward(
[2024-05-10T11:03:33.715Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/spade_diffusion_model_unet.py", line 922, in forward
[2024-05-10T11:03:33.715Z]     h = self.middle_block(hidden_states=h, temb=emb, context=context)
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.715Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.715Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/diffusion_model_unet.py", line 793, in forward
[2024-05-10T11:03:33.715Z]     hidden_states = self.attention(hidden_states)
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.715Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.715Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/blocks/spatialattention.py", line 76, in forward
[2024-05-10T11:03:33.715Z]     x = self.norm(x)
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.715Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/normalization.py", line 268, in forward
[2024-05-10T11:03:33.715Z]     return F.group_norm(
[2024-05-10T11:03:33.715Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/functional.py", line 2360, in group_norm
[2024-05-10T11:03:33.715Z]     return torch.group_norm(input, num_groups, weight, bias, eps, torch.backends.cudnn.enabled)
[2024-05-10T11:03:33.715Z] RuntimeError: Unsupported memory format. Supports only ChannelsLast, Contiguous
[2024-05-10T11:03:33.715Z] 
[2024-05-10T11:03:33.715Z] ======================================================================
[2024-05-10T11:03:33.715Z] ERROR: test_shape_with_different_in_channel_out_channel (tests.test_spade_diffusion_model_unet.TestDiffusionModelUNet3D)
[2024-05-10T11:03:33.715Z] ----------------------------------------------------------------------
[2024-05-10T11:03:33.971Z] 
[2024-05-10T11:03:33.971Z] 
[2024-05-10T11:03:33.971Z] tests finished, printing completed times >10.0s in ascending order...
[2024-05-10T11:03:33.971Z] 
[2024-05-10T11:03:33.971Z] test_start (tests.test_loader_semaphore.TestImportLock) (11.9s)
[2024-05-10T11:03:33.971Z] test_visible_devices (tests.test_set_visible_devices.TestVisibleDevices) (12.0s)
[2024-05-10T11:03:33.971Z] test_pdb (tests.test_config_parser.TestConfigParser) (12.1s)
[2024-05-10T11:03:33.971Z] test_verify_fp16_0__home_jenkins_agent_workspace_MONAI_premerge_monai_tests_testing_data_metadata_json (tests.test_bundle_verify_net.TestVerifyNetwork) (12.9s)
[2024-05-10T11:03:33.971Z] test_densenet_shape_03 (tests.test_densenet.TestDENSENET) (13.4s)
[2024-05-10T11:03:33.971Z] test_script (tests.test_highresnet.TestHighResNet) (15.6s)
[2024-05-10T11:03:33.971Z] test_invert (tests.test_invertd.TestInvertd) (16.0s)
[2024-05-10T11:03:33.971Z] test_cpu_precise_backwards_5_3_dimensions_1_channel_high_spatial_sigmas_high_color_sigma (tests.test_trainable_bilateral.BilateralFilterTestCaseCpuPrecise) (16.0s)
[2024-05-10T11:03:33.971Z] test_lr_finder (tests.test_lr_finder.TestLRFinder) (28.2s)
[2024-05-10T11:03:33.971Z] test discovery time: 15.0s
[2024-05-10T11:03:33.971Z] total testing time: 1.38e+02s
[2024-05-10T11:03:33.971Z] Remember to check above times for any errors!
[2024-05-10T11:03:33.971Z] Traceback (most recent call last):
[2024-05-10T11:03:33.971Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_spade_diffusion_model_unet.py", line 541, in test_shape_with_different_in_channel_out_channel
[2024-05-10T11:03:33.971Z]     result = net.forward(
[2024-05-10T11:03:33.971Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/spade_diffusion_model_unet.py", line 922, in forward
[2024-05-10T11:03:33.971Z]     h = self.middle_block(hidden_states=h, temb=emb, context=context)
[2024-05-10T11:03:33.971Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.971Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.971Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/nets/diffusion_model_unet.py", line 793, in forward
[2024-05-10T11:03:33.971Z]     hidden_states = self.attention(hidden_states)
[2024-05-10T11:03:33.971Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.971Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.971Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/networks/blocks/spatialattention.py", line 76, in forward
[2024-05-10T11:03:33.971Z]     x = self.norm(x)
[2024-05-10T11:03:33.971Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
[2024-05-10T11:03:33.971Z]     return forward_call(*input, **kwargs)
[2024-05-10T11:03:33.971Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/normalization.py", line 268, in forward
[2024-05-10T11:03:33.971Z]     return F.group_norm(
[2024-05-10T11:03:33.971Z]   File "/usr/local/lib/python3.8/dist-packages/torch/nn/functional.py", line 2360, in group_norm
[2024-05-10T11:03:33.971Z]     return torch.group_norm(input, num_groups, weight, bias, eps, torch.backends.cudnn.enabled)
[2024-05-10T11:03:33.971Z] RuntimeError: Unsupported memory format. Supports only ChannelsLast, Contiguous
```

</p>
</details>

marksgraham commented:
sorry i didn't think about the spade version. 🤞 🤞 this  time

KumoLiu commented:
/build

KumoLiu commented:
/build
